src/linear_algebra/exterior_algebra.lean

```diff
@@ -0,0 +1,145 @@
+which is denoted `exterior_algebra.lift R f cond`.
+
+The canonical linear map `M → exterior_algebra R M` is denoted `exterior_algebra.ι R`.
+
+## Theorems
+
+The main theorems proved ensure that `exterior_algebra R M` satisfies the universal property
+of the exterior algebra.
+1. `ι_comp_lift` is  fact that the composition of `ι R` with `lift R f cond` agrees with `f`.
+2. `lift_unique` ensures the uniqueness of `lift R f cond` with respect to 1.
+
+## Implementation details
+
+The exterior algebra of `M` is constructed as a quotient of the tensor algebra, as follows.
+1. We define a relation `exterior_algebra.rel R M` on `tensor_algebra R M`.
+   This is the smallest relation which identifies squares of elements of `M` with `0`.
+2. The exterior algebra is the quotient of the tensor algebra by this relation.
+
+-/
+
+variables (R : Type*) [comm_semiring R]
+variables (M : Type*) [add_comm_monoid M] [semimodule R M]
+
+namespace exterior_algebra
+open tensor_algebra
+
+/--
+An inductively defined relation on `tensor_algebra R M` used to define the exterior algebra.
+-/
+inductive rel : tensor_algebra R M → tensor_algebra R M → Prop
+| of (m : M) : rel ((ι R m) * (ι R m)) 0
+
+end exterior_algebra
+
+/--
+The exterior algebra of an `R`-semimodule `M`.
+-/
+@[derive [inhabited, semiring, algebra R]]
+def exterior_algebra := ring_quot (exterior_algebra.rel R M)
+
+namespace exterior_algebra
+
+variables {M}
+
+/--
+The canonical linear map `M →ₗ[R] exterior_algebra R M`.
+-/
+def ι : M →ₗ[R] exterior_algebra R M :=
+  (ring_quot.mk_alg_hom R _).to_linear_map.comp (tensor_algebra.ι R)
+
+
+variables {R}
+
+/-- As well as being linear, `ι m` squares to zero -/
+@[simp]
+theorem ι_square_zero (m : M) : (ι R m) * (ι R m) = 0 :=
+begin
+  dsimp [ι],
+  rw [←alg_hom.map_mul, ←alg_hom.map_zero _],
+  exact ring_quot.mk_alg_hom_rel R (rel.of m),
+end
+
+variables (R)
+
+/--
+Given a linear map `f : M →ₗ[R] A` into an `R`-algebra `A`, which satisfies the condition:
+`cond : ∀ m : M, f m * f m = 0`, this is the canonical lift of `f` to a morphism of `R`-algebras
+from `exterior_algebra R M` to `A`.
+-/
+def lift {A : Type*} [semiring A] [algebra R A] (f : M →ₗ[R] A) (cond : ∀ m, f m * f m = 0) :
+  exterior_algebra R M →ₐ[R] A :=
+ring_quot.lift_alg_hom R (tensor_algebra.lift R f)
+  (λ x y h, by {
+    induction h,
+    rw [alg_hom.map_zero, alg_hom.map_mul, tensor_algebra.lift_ι_apply, cond],
+  })
+
+@[simp]
+theorem ι_comp_lift {A : Type*} [semiring A] [algebra R A] (f : M →ₗ[R] A)
+(cond : ∀ m, f m * f m = 0) :
+  (lift R f cond).to_linear_map.comp (ι R) = f := by {ext, simp [lift, ι], }
+
+@[simp]
+theorem lift_ι_apply {A : Type*} [semiring A] [algebra R A] (f : M →ₗ[R] A)
+(cond : ∀ m, f m * f m = 0) (x) :
+  lift R f cond (ι R x) = f x := by { dsimp [lift, ι], rw tensor_algebra.lift_ι_apply, }
+
+@[simp]
+theorem lift_unique {A : Type*} [semiring A] [algebra R A] (f : M →ₗ[R] A)
```

Comment:
I see a lot of `rw ← lift_unique`, could you turn it into its own lemma?

Reply:
(And/or consolidate it with `lift_comp_ι`?)

Reply:
Again, the lemma names are all copied from `free_algebra`, and already appear in `universal_enveloping`. Clearly we're missing some abstraction since we keep repeating them, and I'd rather hold off on more lemmas until someone more mathematical than me works out what that abstraction is. cc @semorrison.

Reply:
Hmm, some of the code in `free_algebra` is also not optimal, so I'm not convinced consistency is better than efficiency here.

Presumably, the abstraction will be [some form of universal algebra metaprogram](https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/Free.20stuff).

Reply:
> Hmm, some of the code in free_algebra is also not optimal, so I'm not convinced consistency is better than efficiency here.

If the lemma naming matches between the two files, its much easier for someone who knows what they're doing to clean everything up all in one go.

Reply:
You're right. It's better to get someone to invest some more time into cleaning up everything, than make you spend time on only this.